pkg/controller/job/job_controller.go

```diff
@@ -902,9 +902,19 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 
 	var manageJobErr error
 
+	// This is the starting point for evaluating the end state of the Job.
+	// Note that we need to order evaluations since a Job could satisfy multiple criteria at the same time in some cases:
+	// 1. Evaluate the pre-existing SuccessCriteriaMet and FailureTarget to respect the previous reconcile results.
+	// 2. Evaluate failure scenarios
+	// 3. Evaluate success scenarios
+
 	exceedsBackoffLimit := jobCtx.failed > *job.Spec.BackoffLimit
+	// Evaluate the pre-existing SuccessCriteriaMet.
 	jobCtx.finishedCondition = hasSuccessCriteriaMetCondition(&job)
 
+	// Evaluate the pre-existing FailureTarget.
+	// After that evaluate failure scenarios for the PodFailurePolicy only when the Job does not have FailureTarget.
```

Comment:
nit: I think this sentence is redundant, it also feels inconsistent - we don't add the sentence wrt to SuccessCriteriaMet. So I propose to drop it completely.

Reply:
Actually, there are a slightly different between SuccessCriteriaMet and FailureTarget evaluation like below:

- SuccessCriteriaMet
  - we evaluate the pre-existing one and new one in separate places (pre-existing is Line 912, new one is Line 959)
- FailureTarget
  - we evaluate both in here (Line 920 - 927)

Reply:
So, pre-existing and new one evaluations for FailureTarget are combined to here.

Reply:
I see, in that case, I would suggest to 
1. rephrase. for consistency, the comment "// After that evaluate failure scenarios for the PodFailurePolicy only when the Job does not have FailureTarget." as "Evaluate the failure scenarios for the PodFailurePolicy." 
1. move both comments to L921, just before "failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget)".

Reply:
That makes sense.
Done.